File: cs50/sentimental-credit/credit.py
import sys


def main():
    while True:
        credit = input("Number: ")

        if (credit.isnumeric()):
            break

    length = len(credit)

    if (length != 13 and length != 15 and length != 16):
        print("INVALID")
        sys.exit(0)

    checkSum(credit, length)

    firstDigits = int(credit[0] + credit[1])

    if (int(credit[0]) == 4):
        print("VISA")
    elif (firstDigits >= 51 and firstDigits <= 55):
        print("MASTERCARD")
    elif (firstDigits == 34 or firstDigits == 37):
        print("AMEX")
    else:
        print("INVALID")


def checkSum(credit, length):
    credit = int(credit)
    total = False
    digit1 = 0
    digit2 = 0
    sumEvens = 0
    sumOdds = 0
    count = 0

    while credit > 0:
    	digit2 = digit1
    	digit1 = credit % 10

    	if (count % 2 == 0):
            sumEvens += digit1
    	else:
            multiple = 2 * digit1
            sumOdds += (multiple // 10) + (multiple % 10)

    	credit //= 10
    	count += 1

    total = (sumEvens + sumOdds) % 10 == 0
    if (total == False):
        print("INVALID")
        sys.exit(0)


if __name__ == "__main__":
    main()

File: cs50/credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

void checkSum(long credit, int length);
int getNumber(long credit, int index, int offset);
int getLength(long credit);

int main(void)
{
    long credit;

    // Gets input from the user
    do
    {
        credit = get_long("Number: ");
    }
    while (credit < 0);

    // Gets length of the input
    int length = getLength(credit);

    if (length != 13 && length != 15 && length != 16)
    {
        printf("INVALID\n");
        return 0;
    }

    // Checks if the inputted credit is valid
    checkSum(credit, length);

    // Gets the first digit of the card number
    int firstDigits = getNumber(credit, length, 2);

    if (firstDigits / 10 == 4)
    {
        printf("VISA\n");
    }
    else if (firstDigits >= 51 && firstDigits <= 55)
    {
        printf("MASTERCARD\n");
    }
    else if (firstDigits == 34 || firstDigits == 37)
    {
        printf("AMEX\n");
    }
    else
    {
        printf("INVALID\n");
    }

    return 0;
}

int getLength(long credit)
{
    int length = 0;

    // Counting Length
    while (credit > 0)
    {
        credit = credit / 10;
        length++;
    }

    return length;
}

int getNumber(long credit, int index, int offset)
{
    // Retrieves number at the index
    long firstNum = pow(10, index);
    long secondNum = pow(10, index - offset);
    return (credit % firstNum) / secondNum;
}

void checkSum(long credit, int length)
{
    int sum = 0;

    for (int i = 1; i <= length; i++)
    {
        // Second to last number
        if (i % 2 == 0)
        {
            int number = 2 * getNumber(credit, i, 1);
            sum += number < 10 ? number : 1 + (number % 10);
        }
        // Last number
        else
        {
            sum += getNumber(credit, i, 1);
        }
    }

    if (sum % 10 != 0)
    {
        printf("INVALID\n");
        exit(0);
    }
}
